README
README for AREM 1.0.1, based on MACS 1.4.0rc2
Time-stamp: <2011-03-01 18:21:42 Jake Biesinger>

* Introduction

High-throughput sequencing coupled to chromatin immuno-
precipitation (ChIP-Seq) is widely used in characterizing genome-wide
binding patterns of transcription factors, cofactors, chromatin modifiers,
and other DNA binding proteins. A key step in ChIP-Seq data analysis
is to map short reads from high-throughput sequencing to a reference
genome and identify peak regions enriched with short reads. Although
several methods have been proposed for ChIP-Seq analysis, most ex-
isting methods only consider reads that can be uniquely placed in the
reference genome, and therefore have low power for detecting peaks lo-
cated within repeat sequences. Here we introduce a probabilistic ap-
proach for ChIP-Seq data analysis which utilizes all reads, providing a
truly genome-wide view of binding patterns. Reads are modeled using a
mixture model corresponding to K enriched regions and a null genomic
background. We use maximum likelihood to estimate the locations of the
enriched regions, and implement an expectation-maximization (E-M) al-
gorithm, called AREM, to update the alignment probabilities of each
read to different genomic locations.

For additional information, see our paper in RECOMB 2011 or visit our website:
http://cbcl.ics.uci.edu/AREM

AREM is based on the popular MACS peak caller, as described below:

With the improvement of sequencing techniques, chromatin
immunoprecipitation followed by high throughput sequencing (ChIP-Seq)
is getting popular to study genome-wide protein-DNA interactions. To
address the lack of powerful ChIP-Seq analysis method, we present a
novel algorithm, named Model-based Analysis of ChIP-Seq (MACS), for
identifying transcript factor binding sites. MACS captures the
influence of genome complexity to evaluate the significance of
enriched ChIP regions, and MACS improves the spatial resolution of
binding sites through combining the information of both sequencing tag
position and orientation. MACS can be easily used for ChIP-Seq data
alone, or with control sample with the increase of specificity.

The original MACS package is available at: http://liulab.dfci.harvard.edu/MACS/

* Install

Please check the file 'INSTALL' in the distribution.

* Usage

Usage: arem <-t tfile> [-n name] [-g genomesize] [options]

Example: arem -t ChIP.bam -c Control.bam -f BAM -g h -n test -w --call-subpeaks


arem -- Aligning Reads by Expectation-Maximization, based on Model-based Analysis for ChIP-Sequencing (MACS)

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit.
  -t TFILE, --treatment=TFILE
                        ChIP-seq treatment files. REQUIRED. When ELANDMULTIPET
                        is selected, you must provide two files separated by
                        comma, e.g.
                        s_1_1_eland_multi.txt,s_1_2_eland_multi.txt
  -c CFILE, --control=CFILE
                        Control files. When ELANDMULTIPET is selected, you
                        must provide two files separated by comma, e.g.
                        s_2_1_eland_multi.txt,s_2_2_eland_multi.txt
  -n NAME, --name=NAME  Experiment name, which will be used to generate output
                        file names. DEFAULT: "NA"
  -f FORMAT, --format=FORMAT
                        Format of tag file, "AUTO", "BED" or "ELAND" or
                        "ELANDMULTI" or "ELANDMULTIPET" or "ELANDEXPORT" or
                        "SAM" or "BAM" or "BOWTIE". The default AUTO option
                        will let MACS decide which format the file is. Please
                        check the definition in 00README file if you choose EL
                        AND/ELANDMULTI/ELANDMULTIPET/ELANDEXPORT/SAM/BAM/BOWTI
                        E. DEFAULT: "AUTO"
  --petdist=PETDIST     Best distance between Pair-End Tags. Only available
                        when format is 'ELANDMULTIPET'. DEFAULT: 200
  -g GSIZE, --gsize=GSIZE
                        Effective genome size. It can be 1.0e+9 or 1000000000,
                        or shortcuts:'hs' for human (2.7e9), 'mm' for mouse
                        (1.87e9), 'ce' for C. elegans (9e7) and 'dm' for
                        fruitfly (1.2e8), Default:hs
  -s TSIZE, --tsize=TSIZE
                        Tag size. This will overide the auto detected tag
                        size. DEFAULT: 25
  --bw=BW               Band width. This value is only used while building the
                        shifting model. DEFAULT: 300
  -p PVALUE, --pvalue=PVALUE
                        Pvalue cutoff for peak detection. DEFAULT: 1e-5
  -m MFOLD, --mfold=MFOLD
                        Select the regions within MFOLD range of high-
                        confidence enrichment ratio against background to
                        build model. The regions must be lower than upper
                        limit, and higher than the lower limit. DEFAULT:10,30
  --nolambda            If True, MACS will use fixed background lambda as
                        local lambda for every peak region. Normally, MACS
                        calculates a dynamic local lambda to reflect the local
                        bias due to potential chromatin structure.
  --slocal=SMALLLOCAL   The small nearby region in basepairs to calculate
                        dynamic lambda. This is used to capture the bias near
                        the peak summit region. Invalid if there is no control
                        data. DEFAULT: 1000
  --llocal=LARGELOCAL   The large nearby region in basepairs to calculate
                        dynamic lambda. This is used to capture the surround
                        bias. DEFAULT: 10000.
  --off-auto            Whether turn off the auto pair model process. If not
                        set, when MACS failed to build paired model, it will
                        use the nomodel settings, the '--shiftsize' parameter
                        to shift and extend each tags. DEFAULT: False
  --nomodel             Whether or not to build the shifting model. If True,
                        MACS will not build model. by default it means
                        shifting size = 100, try to set shiftsize to change
                        it. DEFAULT: False
  --shiftsize=SHIFTSIZE
                        The arbitrary shift size in bp. When nomodel is true,
                        MACS will use this value as 1/2 of fragment size.
                        DEFAULT: 100
  --keep-dup=KEEPDUPLICATES
                        It controls the MACS behavior towards duplicate tags
                        at the exact same location -- the same coordination
                        and the same strand. The default 'auto' option makes
                        MACS calculate the maximum tags at the exact same
                        location based on binomal distribution using 1e-5 as
                        pvalue cutoff; and the 'all' option keeps every tags.
                        If an integer is given, at most this number of tags
                        will be kept at the same location. Default: auto
  --to-small            When set, scale the larger dataset down to the smaller
                        dataset, by default, the smaller dataset will be
                        scaled towards the larger dataset. DEFAULT: False
  -w, --wig             Whether or not to save extended fragment pileup at
                        every WIGEXTEND bps into a wiggle file. When --single-
                        profile is on, only one file for the whole genome is
                        saved. WARNING: this process is time/space consuming!!
  -B, --bdg             Whether or not to save extended fragment pileup at
                        every bp into a bedGraph file. When it's on, -w,
                        --space and --call-subpeaks will be ignored. When
                        --single-profile is on, only one file for the whole
                        genome is saved. WARNING: this process is time/space
                        consuming!!
  -S, --single-profile  When set, a single wiggle file will be saved for
                        treatment and input. Default: False
  --space=SPACE         The resoluation for saving wiggle files, by default,
                        MACS will save the raw tag count every 10 bps. Usable
                        only with '--wig' option.
  --call-subpeaks       If set, MACS will invoke Mali Salmon's PeakSplitter
                        soft through system call. If PeakSplitter can't be
                        found, an instruction will be shown for downloading
                        and installing the PeakSplitter package. -w option
                        needs to be on and -B should be off to let it work.
                        DEFAULT: False
  --verbose=VERBOSE     Set verbose level. 0: only show critical message, 1:
                        show additional warning message, 2: show process
                        information, 3: show debug messages. DEFAULT:2
  --diag                Whether or not to produce a diagnosis report. It's up
                        to 9X time consuming. Please check 00README file for
                        detail. DEFAULT: False
  --fe-min=FEMIN        For diagnostics, min fold enrichment to consider.
                        DEFAULT: 0
  --fe-max=FEMAX        For diagnostics, max fold enrichment to consider.
                        DEFAULT: maximum fold enrichment
  --fe-step=FESTEP      For diagnostics, fold enrichment step.  DEFAULT: 20
  --no-EM               Do NOT iteratively align multi-reads by E-M. Multi-
                        read probabilities will be based on quality scores or
                        uniform (if --no-quals) DEFAULT : FALSE
  --EM-converge-diff=MIN_CHANGE
                        The minimum entropy change between iterations before
                        halting E-M steps. DEFAULT : 1e-05
  --EM-min-score=MIN_SCORE
                        Minimum enrichment score. Windows below this threshold
                        will all look the same to the aligner. DEFAULT : 1.5
  --EM-max-score=MAX_SCORE
                        Maximum enrichment score. Windows above this threshold
                        will all look the same to the aligner, DEFAULT : No
                        Maximum
  --EM-show-graphs      generate diagnostic graphs for E-M. (requires
                        MATPLOTLIB). DEFAULT : FALSE
  --quality-scale=QUAL_SCALE
                        Initial alignment probabilities are determined by read
                        quality and mismatches. Each possible alignment is
                        assigned a probability from the product over all bases
                        of either 1-p(ReadError_base) when there is no
                        mismatch, or p(ReadError_base) when the called base
                        disagrees with the reference.  You may also select a
                        uniform initialization. Read quality scale is the must
                        be one of ['auto', 'sanger+33', 'illumina+64'].
                        DEFAULT : auto
  --random-multi        Convert all multi reads to unique reads by selecting
                        one alignment at random for each read. DEFAULT : False
  --no-multi            Throw away all reads that have more than one alignment
  --no-greedy-caller    Use AREM default peak caller instead of the greedy
                        caller. This normally results in wider, less enriched
                        peaks, especially with multi-reads. DEFAULT : False
  --no-map-quals        Do not use mapping probabilities as priors in each
                        update step; just use relative enrichment. DEFAULT :
                        False
  --prior-snp=PRIOR_PROB_SNP
                        Prior probability that a SNP occurs at any base in the
                        genome. DEFAULT : 0.001
  --write-read-probs    Write out all final reads, including their alignment
                        probabilities as a BED file. DEFAULT : FALSE

** Parameters:

*** -t/--treatment FILENAME

This is the only REQUIRED parameter for MACS. If the format is
ELANDMULTIPET, user must provide two treatment files separated by
comma, e.g. s_1_1_eland_multi.txt,s_1_2_eland_multi.txt.

*** -c/--control

The control or mock data file in either BED format or any ELAND output
format specified by --format option. Please follow the same direction
as for -t/--treatment.

*** -n/--name

The name string of the experiment. MACS will use this string NAME to
create output files like 'NAME_peaks.xls', 'NAME_negative_peaks.xls',
'NAME_peaks.bed' , 'NAME_summits.bed', 'NAME_model.r' and so on. So
please avoid any confliction between these filenames and your existing
files.

*** -f/--format FORMAT

Format of tag file, can be "ELAND", "BED", "ELANDMULTI",
"ELANDEXPORT", "ELANDMULTIPET" (for pair-end tags), "SAM", "BAM" or
"BOWTIE". Default is "AUTO" which will allow MACS to decide the format
automatically. Please use "AUTO" only when you combine different
formats of files.

The BED format is defined in "http://genome.ucsc.edu/FAQ/FAQformat#format1".

If the format is ELAND, the file must be ELAND result output file,
each line MUST represents only ONE tag, with fields of:

 1. Sequence name (derived from file name and line number if format is not Fasta)
 2. Sequence
 3. Type of match: 
 NM - no match found.
 QC - no matching done: QC failure (too many Ns basically).
 RM - no matching done: repeat masked (may be seen if repeatFile.txt was specified).
 U0 - Best match found was a unique exact match.
 U1 - Best match found was a unique 1-error match. 
 U2 - Best match found was a unique 2-error match. 
 R0 - Multiple exact matches found.
 R1 - Multiple 1-error matches found, no exact matches.
 R2 - Multiple 2-error matches found, no exact or 1-error matches.
 4. Number of exact matches found.
 5. Number of 1-error matches found.
 6. Number of 2-error matches found.
 Rest of fields are only seen if a unique best match was found (i.e. the match code in field 3 begins with "U").
 7. Genome file in which match was found.
 8. Position of match (bases in file are numbered starting at 1).
 9. Direction of match (F=forward strand, R=reverse).
 10. How N characters in read were interpreted: ("."=not applicable, "D"=deletion, "I"=insertion).
 Rest of fields are only seen in the case of a unique inexact match (i.e. the match code was U1 or U2).
 11. Position and type of first substitution error (e.g. 12A: base 12 was A, not whatever is was in read).
 12. Position and type of first substitution error, as above. 

If the format is ELANDMULTI, the file must be ELAND output file from
multiple-match mode, each line MUST represents only ONE tag, with
fields of:

 1. Sequence name 

ChangeLog
2011-03-01  Jake Biesinger  <jake.biesinger@gmail.com>
	AREM version 1.0.1
	
	* Merged AREM with MACS rc2
	
	* Fixed BED parser
	
	* Multireads can now be read in using the SAM and BAM format, in addition
	to the usual BOWTIE format. Please use the provided multiread filtering
	script (multiread_filter) to remove redundant multi-tags.

2011-02-21  Tao Liu  <taoliu@jimmy.harvard.edu>
	Version 1.4.0rc2 (Valentine)

	* --single-wig option is renamed to --single-profile

	* BedGraph output with --bdg or -B option.

	The BedGraph output provides 1bp resolution fragment pileup
	profile. File size is smaller than wig file. This option can be
	combined with --single-profile option to produce a bedgraph file
	for the whole genome. This option can also make --space,
	--call-subpeaks invalid.

	* Fix the description of --shiftsize to correctly state that the
	value is 1/2 d (fragment size).

	* Fix a bug in the call to __filter_w_control_tags when control is
	not available.

	* Fix a bug on --to-small option. Now it works as expected.

	* Fix a bug while counting the tags in candidate peak region, an
	extra tag may be included. (Thanks to Jake Biesinger!)

	* Fix the bug for the peaks extended outside of chromosome
	start. If the minus strand tag goes outside of chromosome start
	after extension of d, it will be thrown out.

	* Post-process script for a combined wig file:

	The "wignorm" command can be called after a full run of MACS14 as
	a postprocess. wignorm can calculate the local background from the
	control wig file from MACS14, then use either foldchange,
	-10*log10(pvalue) from possion test, or difference after asinh
	transformation as the score to build a single wig track to
	represent the binding strength. This script will take a
	significant long time to process.

	* --wigextend has been obsoleted.

2011-01-20 Jake Biesinger <jake.biesinger@gmail.com>
	AREM version 1.0.0, based on MACS version 1.4.0beta (summer wishes)
	
	This version introduces some major modifications to MACS, including new
	packaging as ``AREM'': Aligning Reads By Expectation-Maximization. Most
	importantly, we've added functionality to iteratively align reads with
	ambiguous mappings.
	
	* New Features
	
	** Aligning Reads by Expectation-Maximization
	
	After the tag shift and candidate peak calling phases, multi-reads are
	iteratively aligned. Regions that are enriched (vs. control) are
	considered more likely to be the true genomic source of the multi-reads.
	The method also takes mapping qualities into account.
	
	** Multi-reads: accepted formats
	
	SAM, BAM, and BOWTIE formats recognize multi-reads (several alignments
	for the same read). To be grouped properly, the tag file must be
	sorted by tag name. We've tested using up to 80 alignments per read.
	
	** Multi-reads: quality scores
	
	MAPQ scores will be used directly (if provided by the aligner, e.g., MAQ)
	or calculated based on sequencing read quality and mismatch info. Also,
	there is an option to ignore MAPQ entirely, so that alignment probabilities
	are based solely on the relative enrichment of the possible alignments.
	
	** Multi-reads: chooose, reject, no EM
	
	Options to choose one of the set of possible multi-reads at random, throw
	all multi-reads out altogether, or split the alignment probability
	uniformly across all reads and do not perform EM.
	
	** Poisson scoring function for decimal values
	
	To deal with real-valued read probabilities, we use a linear interpolation
	of the (discrete) poisson CDF. 
	
	** "Smooth Greedy" peak refinement
	
	Candidate peaks regions are inspected and an attempt at identifying
	enriched subpeaks is made. Reads are added to the subpeak only if they
	improve the enrichment p-value.  This should lead to a lower FDR, better
	splitting of subpeaks, and peaks with higher resolution.

2010-09-21  Tao Liu  <taoliu@jimmy.harvard.edu>
	Version 1.4.0rc1 (Starry Sky)

	* Duplicate reads option

	--keep-dup behavior is changed. Now user can specify how many
	reads he/she wants to keep at the same genomic location. 'auto' to
	let MACS decide the number based on binomial distribution, 'all'
	to let MACS keep all reads.

	* pvalue and FDR fixes (Thanks to Prof. Zhiping Weng)

	By default, MACS will now scale the smaller dataset to the bigger
	dataset. For instance, if IP has 10 million reads, and Input has 5
	million, MACS will double the lambda value calculated from Input
	reads while calling BOTH the positive peaks and negative
	peaks. This will address the issue caused by unbalanced numbers of
	reads from IP and Input. If --to-small is turned on, MACS will
	scale the larger dataset to the smaller one. So from now on, if d
	is fixed, then the peaks from a MACS call for A vs B should be
	identical to the negative peaks from a B vs A.
>>>>>>> merge_MACS1.4rc2

2010-09-01  Tao Liu  <taoliu@jimmy.harvard.edu>
	Version 1.4.0beta (summer wishes)

	* New features

	** Model building

	The default behavior in the model building step is slightly
	changed. When MACS can't find enough pairs to build model
	(implemented in alpha version) or the modeled fragment length is
	less than 2 times of tag length (implemented in beta version),
	MACS will use 2 times of --shiftsize value as fragment length in
	the later analysis. --off-auto can turn off this default behavior.

	** Redundant tag filtering

	The IO module is rewritten. The redundant tag filtering process
	becomes simpler and works as promise. The maximum allowed number
	of tags at the exact same location is calculated from the
	sequencing depth and genome size using a binomial distribution,
	for both TREAMENT and CONTROL separately. ( previously only
	TREATMENT is considered ) The exact same location means the same
	coordination and the same strand. Then MACS will only keep at most
	this number of tags at the exact same location in the following
	analysis. An option --keep-dup can let MACS skip the filtering and
	keep all the tags. However this may bring in a lot of sequencing

AREM/PeakModel.py
# Time-stamp: <2011-03-01 18:21:42 Jake Biesinger>

"""Description: Emprical model for peak lengths

Copyright (c) 2008,2009 Yong Zhang, Tao Liu <taoliu@jimmy.harvard.edu>
Copyright (c) 2010,2011 Tao Liu <taoliu@jimmy.harvard.edu>

This code is free software; you can redistribute it and/or modify it
under the terms of the Artistic License (see the file COPYING included
with the distribution).

@status: beta
@version: $Revision$
@originalauthor:  Yong Zhang, Tao Liu
@originalcontact: taoliu@jimmy.harvard.edu

Modifications to probabilistically align reads to regions with highest
enrichment performed by Jacob Biesinger. Repackaged as "AREM" in accordance
with copyright restrictions.

@author: Jacob Biesinger, Daniel Newkirk, Alvin Chon 
@contact: jake.biesinger@gmail.com, dnewkirk@uci.edu, achon@uci.edu


Changes to this file since original release of MACS 1.4 (summer wishes):
  December/January 2011
    * Updated names (AREM, not MACS14)
    * Use alignment probabilities for multi-reads
    * Exclude multi-reads from emprical modeling
"""
import sys, time, random

def median (nums):
    """Calculate Median.

    Parameters:
    nums:  list of numbers
    Return Value:
    median value
    """
    p = sorted(nums)
    l = len(p)
    if l%2 == 0:
        return (p[l/2]+p[l/2-1])/2
    else:
        return p[l/2]

class NotEnoughPairsException(Exception):
    def __init__ (self,value):
        self.value = value
    def __str__ (self):
        return repr(self.value)

class PeakModel:
    """Peak Model class.
    """
    def __init__ (self, opt=None, treatment=None, max_pairnum=500, gz = 0, umfold=30, lmfold=10, bw=200, ts = 25, bg=0):
        self.treatment = treatment
        if opt:
            self.gz = opt.gsize
            self.umfold = opt.umfold
            self.lmfold = opt.lmfold
            self.tsize = opt.tsize
            self.bw = opt.bw
            self.info  = opt.info
            self.debug = opt.debug
            self.warn  = opt.warn
            self.error = opt.warn
        else:
            self.gz = gz
            self.umfold = umfold
            self.lmfold = lmfold            
            self.tsize = ts
            self.bg = bg
            self.bw = bw
            self.info  = lambda x: sys.stderr.write(x+"\n")
            self.debug = lambda x: sys.stderr.write(x+"\n")
            self.warn  = lambda x: sys.stderr.write(x+"\n")
            self.error = lambda x: sys.stderr.write(x+"\n")

        self.max_pairnum = max_pairnum
        self.summary = ""
        self.plus_line = None
        self.minus_line = None
        self.shifted_line = None
        self.d = None
        self.scan_window = None
        self.min_tags = None
        self.peaksize = None
        self.build()
    
    def build (self):
        """Build the model.

        prepare self.d, self.scan_window, self.plus_line,
        self.minus_line and self.shifted_line to use.
        """
        self.peaksize = 2*self.bw
        self.min_tags = float(self.treatment.total) * self.lmfold * self.peaksize / self.gz /2 # mininum unique hits on single strand
        self.max_tags = float(self.treatment.total) * self.umfold * self.peaksize / self.gz /2 # maximum unique hits on single strand
        self.debug("#2 tags required in model: %.2f min, %.2f max" % (self.min_tags, self.max_tags))
        # use treatment data to build model
        paired_peakpos = self._paired_peaks ()
        # select up to 1000 pairs of peaks to build model
        # Jake - odd that he selects the first 1000 rather than at random.
        num_paired_peakpos = 0
        num_paired_peakpos_remained = self.max_pairnum
        num_paired_peakpos_picked = 0
        for c in paired_peakpos.keys():
            num_paired_peakpos +=len(paired_peakpos[c])
            if num_paired_peakpos_remained == 0:
                paired_peakpos.pop(c)
            else:
                paired_peakpos[c] = paired_peakpos[c][:num_paired_peakpos_remained]
                num_paired_peakpos_remained -=  len(paired_peakpos[c])
                num_paired_peakpos_picked += len(paired_peakpos[c])

        self.info("#2 number of paired peaks: %d" % (num_paired_peakpos))
        if num_paired_peakpos < 100:
            self.error("Too few paired peaks (%d) so I can not build the model! Broader your MFOLD range parameter may erase this error. If it still can't build the model, please use --nomodel and --shiftsize 100 instead." % (num_paired_peakpos))
            self.error("Process for pairing-model is terminated!")
            raise NotEnoughPairsException("No enough pairs to build model")
        elif num_paired_peakpos < self.max_pairnum:
            self.warn("Fewer paired peaks (%d) than %d! Model may not be build well! Lower your MFOLD parameter may erase this warning. Now I will use %d pairs to build model!" % (num_paired_peakpos,self.max_pairnum,num_paired_peakpos_picked))
        self.debug("Use %d pairs to build the model." % (num_paired_peakpos_picked))
        self._paired_peak_model(paired_peakpos)

    def __str__ (self):
        """For debug...

        """
        return """
Summary of Peak Model:
  Baseline: %d
  Upperline: %d
  Fragment size: %d
  Scan window size: %d
""" % (self.min_tags,self.max_tags,self.d,self.scan_window)

    def _paired_peak_model (self, paired_peakpos):
        """Use paired peak positions and treatment tag positions to build the model.

        Modify self.(d, model_shift size and scan_window size. and extra, plus_line, minus_line and shifted_line for plotting).
        """
        window_size = 1+2*self.peaksize
        self.plus_line = [0]*window_size
        self.minus_line = [0]*window_size
        for chrom in paired_peakpos.keys():
            paired_peakpos_chrom = paired_peakpos[chrom]
            tags_plus, tags_minus = self.treatment.get_locations_by_chr(chrom)
            index_plus, index_minus = self.treatment.get_indexes_by_chr(chrom)
            # every paired peak has plus line and minus line
            #  add plus_line
            self.plus_line = self._model_add_line (paired_peakpos_chrom, tags_plus, index_plus, self.plus_line)
            #  add minus_line
            self.minus_line = self._model_add_line (paired_peakpos_chrom, tags_minus, index_minus, self.minus_line)

        # find top 
        plus_tops = []
        minus_tops = []
        plus_max = max(self.plus_line)
        minus_max = max(self.minus_line)
        for i in range(window_size):
            if self.plus_line[i] == plus_max:
                plus_tops.append(i)
            if self.minus_line[i] == minus_max:
                minus_tops.append(i)
        self.d = minus_tops[len(minus_tops)/2] - plus_tops[len(plus_tops)/2] + 1
        print 'plus_tops: %s\nminus_tops: %s\nd: %s\nwindow_size: %s' %(plus_tops, minus_tops, self.d, window_size)
        shift_size = self.d/2
        # find the median point
        #plus_median = median(self.plus_line) 
        #minus_median = median(self.minus_line)       

        self.scan_window = max(self.d,self.tsize)*2
        # a shifted model
        self.shifted_line = [0]*window_size
        plus_shifted = [0]*shift_size
        plus_shifted.extend(self.plus_line[:-1*shift_size])
        minus_shifted = self.minus_line[shift_size:]
        minus_shifted.extend([0]*shift_size)
        for i in range(window_size):
            self.shifted_line[i]=minus_shifted[i]+plus_shifted[i]
        return True

    def _model_add_line (self, pos1, pos2, index2, line):
        """Project each pos in pos2 which is included in
        [pos1-self.peaksize,pos1+self.peaksize] to the line.

        """
        i1 = 0                  # index for pos1
        i2 = 0                  # index for pos2
        i2_prev = 0             # index for pos2 in previous pos1
                                # [pos1-self.peaksize,pos1+self.peaksize]
                                # region
        i1_max = len(pos1)
        i2_max = len(pos2)
        last_p2 = -1
        flag_find_overlap = False
        while i1<i1_max and i2<i2_max:
            p1 = pos1[i1]
            p2 = pos2[i2]
            if p1-self.peaksize > p2 or index2[i2] != 0: # move pos2, skip multi-aligning reads
                i2 += 1
            elif p1+self.peaksize < p2: # move pos1
                i1 += 1                 
                i2 = i2_prev    # search minus peaks from previous index
                flag_find_overlap = False
            else:               # overlap!
                if not flag_find_overlap:
                    flag_find_overlap = True
                    i2_prev = i2 # only the first index is recorded
                # project
                for i in range(p2-p1+self.peaksize-self.tsize/2,p2-p1+self.peaksize+self.tsize/2):
                    if i>=0 and i<len(line):
                        line[i]+=1
                i2+=1
        return line

    def _paired_peaks (self):
        """Call paired peaks from fwtrackI object.

        Return paired peaks center positions.
        """
        chrs = self.treatment.get_chr_names()
        chrs.sort()
        paired_peaks_pos = {}
        for chrom in chrs:
            self.debug("Chromosome: %s" % (chrom))
            tags = self.treatment.get_locations_by_chr(chrom)
            indexes = self.treatment.get_indexes_by_chr(chrom)
            plus_peaksinfo = self._naive_find_peaks (tags[0], indexes[0])
            self.debug("Number of unique tags on + strand: %d" % (len(tags[0])))            
            self.debug("Number of peaks in + strand: %d" % (len(plus_peaksinfo)))
            minus_peaksinfo = self._naive_find_peaks (tags[1], indexes[1])
            self.debug("Number of unique tags on - strand: %d" % (len(tags[1])))            
            self.debug("Number of peaks in - strand: %d" % (len(minus_peaksinfo)))
            if not plus_peaksinfo or not minus_peaksinfo:
                self.debug("Chrom %s is discarded!" % (chrom))
                continue
            else:
                paired_peaks_pos[chrom] = self._find_pair_center (plus_peaksinfo, minus_peaksinfo)
                self.debug("Number of paired peaks: %d" %(len(paired_peaks_pos[chrom])))
        return paired_peaks_pos

    def _find_pair_center (self, pluspeaks, minuspeaks):
        ip = 0                  # index for plus peaks
        im = 0                  # index for minus peaks
        im_prev = 0             # index for minus peaks in previous plus peak
        pair_centers = []
        ip_max = len(pluspeaks)
        im_max = len(minuspeaks)
        flag_find_overlap = False
        while ip<ip_max and im<im_max:
            (pp,pn) = pluspeaks[ip] # for (peakposition, tagnumber in peak)
            (mp,mn) = minuspeaks[im]
            if pp-self.peaksize > mp: # move minus
                im += 1
            elif pp+self.peaksize < mp: # move plus
                ip += 1                 
                im = im_prev    # search minus peaks from previous index
                flag_find_overlap = False
            else:               # overlap!
                if not flag_find_overlap:
                    flag_find_overlap = True
                    im_prev = im # only the first index is recorded
                if float(pn)/mn < 2 and float(pn)/mn > 0.5: # number tags in plus and minus peak region are comparable...
                    if pp < mp:
                        pair_centers.append((pp+mp)/2)
                        #self.debug ( "distance: %d, minus: %d, plus: %d" % (mp-pp,mp,pp))
                im += 1
        return pair_centers
            
    def _naive_find_peaks (self, taglist, indexlist ):
        """Naively call peaks based on tags counting.

        Do NOT include reads with multiple alignments. This exclusion filters
        down through all of the model building phase.

        Return peak positions and the tag number in peak region by a tuple list [(pos,num)].
        """
        peak_info = []    # store peak pos in every peak region and
                          # unique tag number in every peak region
        if len(taglist)<2:
            return peak_info
        
        first_pos = None
        for i in xrange(len(taglist)):
            # get first non-multi read
            if indexlist[i] == 0:
                first_pos = i
                break
        if first_pos is None:
            return peak_info
        current_tag_list = [taglist[first_pos]]
        for i in range(first_pos+1, len(taglist)):
            if indexlist[i] != 0:  # filter out multi reads in building model
                continue
            pos = taglist[i]
            
            # Jake - This step seems incorrect. a stretch of bw*2 is considered,
            # always originating at the next read. Say the stretch divides a
            # peak in half and is most potent in the middle.  Then the counts 
            # will be high at the end of this peak and the beginning of the next.
            # ** one solution would be to do normal peak calling and project the 
            # tags down to the line, looking for contiguous reads that pass 
            # the mfold params.
            # ** That or keep the bw*2 size, but use the smooth-greedy method to
            # put the bw in the center of the peak.
            # ** I think it's wise to keep the multi-reads out of this process. 
            # otherwise, we might end up modeling the lengths of some transposons!
            if (pos-current_tag_list[0]+1) > self.peaksize: # call peak in current_tag_list
                # a peak will be called if tag number is ge min tags.
                if len(current_tag_list) >= self.min_tags and len(current_tag_list) <= self.max_tags:
                    peak_info.append((self._naive_peak_pos(current_tag_list),len(current_tag_list)))
                current_tag_list = [] # reset current_tag_list

            current_tag_list.append(pos)   # add pos while 1. no
                                           # need to call peak;
                                           # 2. current_tag_list is []
        return peak_info

    def _naive_peak_pos (self, pos_list ):
        """Naively calculate the position of peak.

        return the highest peak summit position.
        """
        peak_length = pos_list[-1]+1-pos_list[0]+self.tsize
        start = pos_list[0] -self.tsize/2
        horizon_line = [0]*peak_length # the line for tags to be projected
        for pos in pos_list:
            for pp in range(int(pos-start-self.tsize/2),int(pos-start+self.tsize/2)): # projected point
                horizon_line[pp] += 1

        top_pos = []            # to record the top positions. Maybe > 1
        top_p_num = 0           # the maximum number of projected points
        for pp in range(peak_length): # find the peak posistion as the highest point
            if horizon_line[pp] > top_p_num:
                top_p_num = horizon_line[pp]
                top_pos = [pp]
            elif horizon_line[pp] == top_p_num:
                top_pos.append(pp)
        return (top_pos[int(len(top_pos)/2)]+start)
